game/fs.cc

```diff
@@ -32,46 +32,218 @@ namespace {
 	}
 
 	const fs::path performous = "performous";
-	const fs::path configSchema = "config/schema.xml";	
-
-	struct PathCache {
-		Paths paths;
-		bool didMigrateConfig = false;
-		fs::path base, share, locale, sysConf, home, conf, data, cache;
-		/// Expand a path specifier as a list of actual paths. Expands ~ (home) and DATADIR (Performous search path).
-	Paths pathExpand(fs::path p) {
-			Paths ret;
-			if (pathRootHack(p, "~")) ret.push_back(home / p);
-			else if (pathRootHack(p, "DATADIR")) {
-				// Add all data paths with p appended to them
-				for (auto const& path: paths) ret.push_back(path / p);
+	const fs::path configSchema = "config/schema.xml";
+
+	std::mutex mutex;
+	using Lock = std::lock_guard<std::mutex>;
+
+	void PathCache::pathBootstrap() {
+		if (!base.empty()) return;  // Only bootstrap once
+		// Base (e.g. /usr/local), share (src or installed data files) and locale (built or installed .mo files)
+		{
+			char const* root = getenv("PERFORMOUS_ROOT");
+			switch (Platform::currentOS()) {
+				case Platform::HostOS::OS_WIN:
+					base = fs::canonical(root ? root : execname().parent_path());
+					break;
+				case Platform::HostOS::OS_MAC:
+					if (const auto appFolder{execname().parent_path().parent_path().parent_path()}; appFolder.extension() == ".app") {
```

Comment:
Would be nice to refactor `base` so it's not kinda arbitrary how many `parent_path()` we need to go up...
Same for Windows and the default case.

Reply:
> Would be nice to refactor `base` so it's not kinda arbitrary how many `parent_path()` we need to go up... Same for Windows and the default case.

Not much can be done about this. Mac apps have the `whatever.app/Contents/MacOS/whatever-exe` structure and that can't really be changed.

The real base I guess would be `Contents` but IMO that looks ugly. In windows, I guess perhaps we could improve it. Also, the purpose here is for it to play nice with something I did in the SpdLog PR:

```c++
std::string formatPath(const fs::path& target) {
	// Normalize the paths by resolving symlinks and removing redundant elements.
	// But, if just a filename, keep it as is.
	fs::path canonicalTarget = target.has_parent_path() ? fs::weakly_canonical(target) : target;
	if (canonicalTarget == cache.base) {
	// Return the absolute path if referring to the base directory.
		return cache.base.string();
	}
	// Check if the target is a descendant of base.
	if (std::mismatch(cache.base.begin(), cache.base.end(), canonicalTarget.begin()).first == cache.base.end()) {
		// If target is a descendant of base, return the relative path starting from base's parent.
		return fs::relative(canonicalTarget, cache.base.parent_path()).string();
	} else {
		// Otherwise, return the absolute path to target.
		return canonicalTarget.string();
	}
}
```